File: appointment/src/main/kotlin/it/beaesthetic/appointment/agenda/domain/event/AgendaEvent.kt
package it.beaesthetic.appointment.agenda.domain.event

import it.beaesthetic.appointment.agenda.domain.Clock
import it.beaesthetic.appointment.agenda.domain.reminder.Reminder
import it.beaesthetic.appointment.agenda.domain.reminder.ReminderStatus
import it.beaesthetic.appointment.common.DomainEventRegistry
import java.time.Duration
import java.time.Instant

@JvmInline value class AgendaEventId(val value: String)

data class Attendee(val id: String, val displayName: String)

data class AgendaEvent(
    val id: AgendaEventId,
    val timeSpan: TimeSpan,
    val attendee: Attendee,
    val cancelReason: CancelReason?,
    val data: AgendaEventData,
    val reminder: Reminder,
    val remindBefore: Duration,
    val createdAt: Instant,
    private val domainEventRegistry: DomainEventRegistry<AgendaLifecycleEvent>
) : DomainEventRegistry<AgendaLifecycleEvent> by domainEventRegistry {

    companion object {
        fun create(
            id: AgendaEventId,
            timeSpan: TimeSpan,
            attendee: Attendee,
            data: AgendaEventData,
            reminderBefore: Duration,
        ): AgendaEvent {
            val agendaEvent =
                AgendaEvent(
                    id,
                    timeSpan,
                    attendee,
                    null,
                    data,
                    Reminder(id, ReminderStatus.PENDING, null),
                    reminderBefore,
                    Instant.now(),
                    DomainEventRegistry.delegate()
                )
            return agendaEvent.also { it.addEvent(AgendaEventScheduled(it)) }
        }
    }

    val title =
        when (data) {
            is BasicEventData -> data.title
            is AppointmentEventData -> data.services.joinToString(",") { it.name }
        }

    fun reschedule(timeSpan: TimeSpan): AgendaEvent {
        require(cancelReason == null) { "Event already canceled" }
        return copy(timeSpan = timeSpan).also { it.addEvent(AgendaEventRescheduled(it)) }
    }

    fun cancel(reason: CancelReason): AgendaEvent {
        require(cancelReason == null) { "Event already canceled" }
        return copy(cancelReason = reason).also { it.addEvent(AgendaEventDeleted(it)) }
    }

    fun updateReminder(reminder: Reminder): AgendaEvent {
        return copy(reminder = reminder)
    }

    fun trackReminderAsSent(clock: Clock): AgendaEvent {
        return copy(reminder = reminder.copy(status = ReminderStatus.SENT, sentAt = clock.now()))
    }
}

File: appointment/internal/domain/agenda_event.go
package domain

import "time"

type AgendaEvent struct {
	ID             string
	Type           EventType
	Title          string
	Description    string
	Start          time.Time
	End            time.Time
	Attendee       Attendee
	Services       []AppointmentServiceRef
	CancelReason   *CancelReason
	ReminderStatus ReminderStatus
	ReminderSentAt *time.Time
	RemindBefore   time.Duration
	Version        int64
	CreatedAt      time.Time
	UpdatedAt      time.Time
	events         []LifecycleEvent
}

func NewAgendaEvent(
	id string,
	typ EventType,
	title string,
	description string,
	start time.Time,
	end time.Time,
	attendee Attendee,
	services []AppointmentServiceRef,
	remindBefore time.Duration,
	now time.Time,
) (AgendaEvent, error) {
	if id == "" || attendee.ID == "" {
		return AgendaEvent{}, ErrMissingRequiredAgendaData
	}
	if !end.After(start) {
		return AgendaEvent{}, ErrInvalidTimeSpan
	}

	event := AgendaEvent{
		ID:             id,
		Type:           typ,
		Title:          title,
		Description:    description,
		Start:          start.UTC(),
		End:            end.UTC(),
		Attendee:       attendee,
		Services:       services,
		ReminderStatus: ReminderPending,
		RemindBefore:   remindBefore,
		CreatedAt:      now.UTC(),
		UpdatedAt:      now.UTC(),
	}
	event.record(AgendaEventScheduled(event.ID))
	return event, nil
}

func (event *AgendaEvent) Reschedule(start time.Time, end time.Time) error {
	if !end.After(start) {
		return ErrInvalidTimeSpan
	}
	event.Start = start.UTC()
	event.End = end.UTC()
	event.UpdatedAt = time.Now().UTC()
	event.record(AgendaEventRescheduled(event.ID))
	return nil
}

func (event *AgendaEvent) Cancel(reason CancelReason) {
	event.CancelReason = &reason
	event.UpdatedAt = time.Now().UTC()
	event.record(AgendaEventDeleted(event.ID))
}

func (event *AgendaEvent) PullEvents() []LifecycleEvent {
	events := event.events
	event.events = nil
	return events
}

func (event *AgendaEvent) record(lifecycleEvent LifecycleEvent) {
	event.events = append(event.events, lifecycleEvent)
}

func (event *AgendaEvent) MarkReminderSentRequested(now time.Time) {
	event.ReminderStatus = ReminderSentRequested
	event.UpdatedAt = now.UTC()
}

func (event *AgendaEvent) MarkReminderSent(now time.Time) {
	event.ReminderStatus = ReminderSent
	sentAt := now.UTC()
	event.ReminderSentAt = &sentAt
	event.UpdatedAt = sentAt
}
